yb-voyager/cmd/common.go

```diff
@@ -479,17 +480,68 @@ func initMetaDB(migrationExportDir string) *metadb.MetaDB {
-		userFacingMsg := fmt.Sprintf("Voyager requires the entire migration workflow to be executed using a single Voyager version.\n"+
-			"The export-dir %q was created using version %q and the current version is %q. Either use Voyager %q to continue the migration or start afresh "+
-			"with a new export-dir.", migrationExportDir, msr.VoyagerVersion, utils.YB_VOYAGER_VERSION, msr.VoyagerVersion)
+
+	// If the msr VoyagerVersion is less than the PREVIOUS_BREAKING_CHANGE_VERSION, then the export-dir is not compatible with the current Voyager version.
+	// This version will always be a final release version and never "main" or "rc" version.
+	previousBreakingChangeVersion, err := version.NewVersion(utils.PREVIOUS_BREAKING_CHANGE_VERSION)
+	if err != nil {
+		utils.ErrExit("could not create version from %q: %v", utils.PREVIOUS_BREAKING_CHANGE_VERSION, err)
+	}
+
+	var versionCheckFailed bool
+
+	if msr.VoyagerVersion == "main" {
+		// If the export-dir was created using the main branch, then the current version should also be the main branch.
+		if utils.YB_VOYAGER_VERSION != "main" {
+			versionCheckFailed = true
+		}
+	} else if msr.VoyagerVersion != "" {
+		msrVoyagerVersionString := msr.VoyagerVersion
+		if strings.Contains(msrVoyagerVersionString, "rc") {
+			// RC version will be like 0rc1.1.8.6
+			// We need to extract 1.8.6 from it
+			// Compring with this 1.8.6 should be enough to check if the version is compatible with the current version
+			// Split the string at "rc" to isolate the part after it
+			parts := strings.Split(msrVoyagerVersionString, "rc")
+			if len(parts) > 1 {
+				// Further split the remaining part by '.' and remove the first segment
+				versionParts := strings.Split(parts[1], ".")
+				if len(versionParts) > 1 {
+					msrVoyagerVersionString = strings.Join(versionParts[1:], ".") // Join the parts after the first one
+				} else {
+					log.Printf("Unexpected version format after 'rc': %s", msrVoyagerVersionString)
+					utils.ErrExit("could not extract version from %q", msr.VoyagerVersion)
+				}
+			} else {
+				log.Printf("Unexpected version format: %s", msrVoyagerVersionString)
+				utils.ErrExit("could not extract version from %q", msr.VoyagerVersion)
+			}
+		}
+
+		msrVoyagerVersion, err := version.NewVersion(msrVoyagerVersionString)
+		if err != nil {
+			utils.ErrExit("could not create version from %q: %v", msr.VoyagerVersion, err)
+		}
+
+		if msrVoyagerVersion.LessThan(previousBreakingChangeVersion) {
+			versionCheckFailed = true
+		}
+	}
+
+	if versionCheckFailed {
+		userFacingMsg := fmt.Sprintf("\nThe export-dir %q was created using voyager version %q. "+
+			"However, the current version %q requires the export-dir to be created using version %q or later. "+
+			"Either use a compatible version to continue the migration or start afresh with a new export-dir. ",
+			migrationExportDir, msr.VoyagerVersion, utils.YB_VOYAGER_VERSION, utils.PREVIOUS_BREAKING_CHANGE_VERSION)
```

Comment:
should we display the `msrVoyagerFinalVersion` instead of `msr.VoyagerVersion` in the display to avoid RC cases.?

Reply:
I think in this case we can display the rc version itself. Since the message states that you created the export dir using this particular version

Reply:
> The export-dir "/home/ubuntu/pg_export2" was created using voyager version "0rc1.1.8.4". However, the current version "main" requires the export-dir to be created using version "1.8.5" or later. Either use a compatible version to continue the migration or start afresh with a new export-dir.

Example

Reply:
Hmm okay